src/server/http/xsrf.js

```diff
@@ -4,18 +4,51 @@ export default function (kbnServer, server, config) {
   const disabled = config.get('server.xsrf.disableProtection');
   const versionHeader = 'kbn-version';
   const xsrfHeader = 'kbn-xsrf';
+  const contentTypeHeader = 'content-type';
+  const allowedRequestMediaTypes = ['application/json', 'application/x-ndjson'];
 
   server.ext('onPostAuth', function (req, reply) {
     if (disabled) {
       return reply.continue();
     }
 
     const isSafeMethod = req.method === 'get' || req.method === 'head';
-    const hasVersionHeader = versionHeader in req.headers;
-    const hasXsrfHeader = xsrfHeader in req.headers;
+    if (!isSafeMethod) {
+      const hasVersionHeader = versionHeader in req.headers;
+      const hasXsrfHeader = xsrfHeader in req.headers;
+      const hasContentTypeHeader = contentTypeHeader in req.headers;
 
-    if (!isSafeMethod && !hasVersionHeader && !hasXsrfHeader) {
-      return reply(badRequest(`Request must contain an ${xsrfHeader} header`));
+      // Since we're only interested in media type let's extract it from the content type ("media type [;parameter]")
+      // and leave off parameter portion (e.g. charset) of the header.
+      const hasAllowedMediaType = hasContentTypeHeader && allowedRequestMediaTypes.includes(
+        req.headers[contentTypeHeader].split(';')[0]
+      );
+
+      if (hasXsrfHeader) {
+        let xsrfHeaderDeprecationMessage =
+          `The ${xsrfHeader} header is deprecated and will be removed in a future version of Kibana.`;
+        if (!hasAllowedMediaType) {
+          xsrfHeaderDeprecationMessage +=
+            ` Specify a ${contentTypeHeader} header of either application/json or application/x-ndjson instead.`;
+        }
+
+        server.log(['warning', 'deprecation'], xsrfHeaderDeprecationMessage);
+      }
+
+      if (!hasAllowedMediaType && hasVersionHeader) {
```

Comment:
Wonder if we should add a `!hasXsrfHeader` to this one. The user is already using the specific xsrf related header and we're logging about it, so this just "adds to the noise" in that case. But